### PR TITLE
[ticket/15007] Add Restart link to installer

### PR DESCRIPTION
**Pull request by zlavdinov**
PHPBB3-15007

Checklist:

- [ ] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [ ] Tests pass
- [ ] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [ ] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-15007

### CODE REVIEW

**Comment by zlavdinov** [comment]
Hello! This PR is for review purposes!

It also seems that I made a mistake and created the branch from master and not from 3.2.x, right?

This fix is a part of me getting to know phpBB. You can see my GSoC student introduction post at area51 forum: https://area51.phpbb.com/phpBB/viewtopic.php?f=99&t=51181&sid=1ef215c06e512a99372e57641f99f271

**Comment by hanakin** [comment]
@zlavdinov you have several testing errors that are failing which are being caused by the naming of your commit msgs. We have strict guidelines for naming and content of commit msgs to ensure proper integration with JIRA. If you follow the last link in the checklist found in your pr msg it will provide you with the corrections that will need to be made for the tests to pass

File: phpBB/adm/style/installer_header.html [hunk 1]
```diff
@@ -48,7 +48,7 @@
 							<li<!-- IF l_block1.S_SELECTED --> id="activemenu"<!-- ENDIF -->><a href="{l_block1.U_TITLE}"><span>{l_block1.L_TITLE}</span></a></li>
 							<!-- END l_block1 -->
 							<!-- BEGIN l_block2 -->
-							<li<!-- IF l_block2.S_SELECTED --> id="activemenu"<!-- ENDIF -->><span<!-- IF l_block2.S_COMPLETE --> class="completed"<!-- ENDIF --> id="installer-stage-{l_block2.STAGE_NAME}">{l_block2.L_TITLE}</span></li>
+							<li<!-- IF l_block2.S_SELECTED --> id="activemenu"<!-- ENDIF -->><span<!-- IF l_block2.S_COMPLETE --> class="completed"<!-- ENDIF --> id="installer-stage-{l_block2.STAGE_NAME}"><!-- IF l_block2.IS_LINK --><a class="{l_block2.CLASS}" href="{l_block2.LINK}" task_index="{l_block2.TASK_INDEX}">{l_block2.L_TITLE}</a><!-- ELSE --> {l_block2.L_TITLE}<!-- ENDIF --> </span></li>
```
Comment (CHItA) [inline phpBB/adm/style/installer_header.html]: I think it should be `data-task-index`

File: phpBB/assets/javascript/installer.js [hunk 2]
```diff
@@ -21,6 +21,26 @@
...
+	// Interception of click on the navigation link
+    $('.my_send').click( function (e) {
+    	e.preventDefault();
+    	// check if the previous fields are filled
+    	if($(this).parent().hasClass('completed')){
+            var xhReq = createXhrObject();
+            xhReq.open('POST', $(this).attr('href'), true);
+            xhReq.setRequestHeader('X-Requested-With', 'XMLHttpRequest');
+            xhReq.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
+            xhReq.send('task_index=' + $(this).attr('task_index'));
```
Comment (CHItA) [inline phpBB/assets/javascript/installer.js]: Same as above

File: phpBB/phpbb/install/helper/navigation/install_navigation.php [hunk 3]
```diff
@@ -58,15 +58,55 @@ public function get()
 						'stage'	=> true,
 						'order'	=> 1,
 					),
-					'obtain_data'	=> array(
-						'label'	=> 'STAGE_OBTAIN_DATA',
+					'obtain_admin_data'	=> array(
+						'label'	=> 'Administrator configuration',
```
Comment (CHItA) [inline phpBB/phpbb/install/helper/navigation/install_navigation.php]: Please don't use hardcoded language strings here.

File: phpBB/phpbb/install/helper/navigation/install_navigation.php [hunk 4]
```diff
@@ -58,15 +58,55 @@ public function get()
...
 						'stage'	=> true,
+                        'is_link'	=> true,
+                        'link'	=> '/install/app.php/install',
+                        'class'	=> 'my_send',
+                        'task_index'	=> '1',
 						'order'	=> 2,
 					),
+                    'obtain_database_data'	=> array(
+                        'label'	=> 'Database configuration',
```
Comment (CHItA) [inline phpBB/phpbb/install/helper/navigation/install_navigation.php]: Hardcoded language

File: phpBB/phpbb/install/helper/navigation/install_navigation.php [hunk 5]
```diff
@@ -58,15 +58,55 @@ public function get()
...
+                        'stage'	=> true,
+                        'is_link'	=> true,
+                        'link'	=> '/install/app.php/install',
+                        'class'	=> 'my_send',
+                        'task_index'	=> '2',
+                        'order'	=> 3,
+                    ),
+                    'obtain_server_data'	=> array(
+                        'label'	=> 'Server configuration',
```
Comment (CHItA) [inline phpBB/phpbb/install/helper/navigation/install_navigation.php]: Hardcoded language string

File: phpBB/phpbb/install/helper/navigation/install_navigation.php [hunk 6]
```diff
@@ -58,15 +58,55 @@ public function get()
...
+                        'stage'	=> true,
+                        'is_link'	=> true,
+                        'link'	=> '/install/app.php/install',
+                        'class'	=> 'my_send',
+                        'task_index'	=> '3',
+                        'order'	=> 4,
+                    ),
+                    'obtain_email_data'	=> array(
+                        'label'	=> 'E-mail configuration',
```
Comment (CHItA) [inline phpBB/phpbb/install/helper/navigation/install_navigation.php]: Hardcoded language string

File: phpBB/phpbb/install/helper/navigation/install_navigation.php [hunk 7]
```diff
@@ -58,15 +58,55 @@ public function get()
...
+                        'stage'	=> true,
+                        'is_link'	=> true,
+                        'link'	=> '/install/app.php/install',
+                        'class'	=> 'my_send',
+                        'task_index'	=> '4',
+                        'order'	=> 5,
+                    ),
+                    'obtain_board_data'	=> array(
+                        'label'	=> 'Bulletin board configuration',
```
Comment (CHItA) [inline phpBB/phpbb/install/helper/navigation/install_navigation.php]: Hardcoded language string

File: phpBB/phpbb/install/installer.php [hunk 8]
```diff
@@ -223,6 +225,10 @@ public function run()
 				}
 
 				$module_index++;
+
```
Comment (CHItA) [inline phpBB/phpbb/install/installer.php]: this should be removed

File: phpBB/phpbb/install/module_base.php [hunk 9]
```diff
@@ -110,10 +110,38 @@ public function run()
 
 		if ($task_index < $iterator->count())
 		{
-			$iterator->seek($task_index);
+		    // Receiving task index
```
Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: I need to look at this when i have some more time, but it seems like something that would increase the execution time possibly quite a bit. 

As far as i understand you try to go back 1 stage with the obtain data stages, which would be fine but i think the ticket refers to restarts after a timeout.

File: phpBB/adm/style/installer_header.html (hunk 1 above)
Comment (zlavdinov) [inline phpBB/adm/style/installer_header.html]: Yes, I will change it. 
I have not finished the work with js, so I will change it further. I do not like how the js works at this point and would like to make more improvements to it

File: phpBB/phpbb/install/helper/navigation/install_navigation.php (hunk 3 above)
Comment (zlavdinov) [inline phpBB/phpbb/install/helper/navigation/install_navigation.php]: Ok, I'll fix it

File: phpBB/phpbb/install/module_base.php (hunk 9 above)
Comment (zlavdinov) [inline phpBB/phpbb/install/module_base.php]: Could you explain what could lead to the increase in the execution time?

So you think that this implementation is incorrect? It seems to me that this is how it should be fixed. I think the user should be able to come back to the previous step.

Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: mostly because of this: https://github.com/phpbb/phpbb/blob/master/phpBB/phpbb/di/service_collection_iterator.php#L44

if it only does this with the data collector services then i think it should be fine. But who knows.

Comment (samatdav) [inline phpBB/phpbb/install/module_base.php]: Would no the increase of the execution time be insignificant since ``` current() ``` is used already many times in similar situations? Moreover, is not it necessary to come to the previous step for this solution? It seems to me that this approach is the only way to return to the previous step.

So you mean that when the user returns to a previous step, the previously inputted information by the user should be auto included back? If so, I agree with the approach and would like to try to implement it. Actually, I already tried to do it, but it turned out to be more difficult and I have not done it yet.

Comment (zlavdinov) [inline phpBB/phpbb/install/module_base.php]: Would no the increase of the execution time be insignificant since ``` current() ``` is used already many times in similar situations? Moreover, is not it necessary to come to the previous step for this solution? It seems to me that this approach is the only way to return to the previous step.

So you mean that when the user returns to a previous step, the previously inputted information by the user should be auto included back? If so, I agree with the approach and would like to try to implement it. Actually, I already tried to do it, but it turned out to be more difficult and I have not done it yet.

File: phpBB/phpbb/install/task_base.php [hunk 10]
```diff
@@ -50,4 +50,10 @@ public function check_requirements()
 	{
 		return true;
 	}
+
+	// Returns the task name for navigation bar
+    public function get_navigation_stage_path()
+    {
+        return array('install', 0, 'instal');
```
Comment (CHItA) [inline phpBB/phpbb/install/task_base.php]: Possible typo there `instal`

File: phpBB/phpbb/install/module_base.php (hunk 9 above)
Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: I'm not sure what you are referring to when you say `current()` being used in similar situations, if you could link me an example of where you seen it, it would be much appreciated. There are other ways to return to whatever point you'd like by setting simply setting the task index.

If you want to retrieve the previously submitted information you should look at the installer's config class. You can also see examples of it being used in other tasks.

File: phpBB/phpbb/install/module_base.php [hunk 11]
```diff
@@ -110,10 +110,38 @@ public function run()
...
+		    $get_task_index = (int)$this->iohandler->get_input('task_index', 0);
+		    // Validation check
+		    if($get_task_index > 0 && $get_task_index < $iterator->count()){
+                $get_task_index--;
+                $iterator->seek($get_task_index);
+                // Setting the menu point as active
+                $this->iohandler->set_active_stage_menu($iterator->current()->get_navigation_stage_path());
```
Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: To be clear, i don't talk about this `current()` call.

File: phpBB/phpbb/install/module_base.php [hunk 12]
```diff
@@ -110,10 +110,38 @@ public function run()
...
+                $this->iohandler->send_response();
+            }
+            else{
+                // Displaying completed as checked
+                for($i = 0; $i < $task_index; $i++){
```
Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: You can use a `foreach` here, it would tidy it up. You instantiate a bunch of tasks just to retrieve an array to send some info to the client side. You already have that information on the client side. The user already clicked a link in a list, making the clicked item active, and every subsequent one inactive. There is not really any reason to do this IMHO.

File: phpBB/phpbb/install/module_base.php (hunk 11 above)
Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: Although the same applies as below, you already had that information client side, why would you want to send it back from the server.

File: phpBB/phpbb/install/module_base.php [hunk 13]
```diff
@@ -110,10 +110,38 @@ public function run()
...
+                $iterator->seek($task_index);
+                // Setting the menu point as active
+                $this->iohandler->set_active_stage_menu($iterator->current()->get_navigation_stage_path());
+                $this->iohandler->send_response();
+            }
 		}
 		else
 		{
+            // Checking all steps as completed
+            for($i = 0; $i < $iterator->count(); $i++){
```
Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: This seems like you can make it visually seem like that a task was completed when it wasn't if you set the task_index to an invalid value. Also doesn't need to be done server side.

**Comment by CHItA** [comment]
This is work in progress I think, so I jumped in a bit early, if I'm not mistaken you can add and remove the WIP label by commenting `!set WIP` and `!unset WIP` respectively. I add the flag for you now.

File: phpBB/phpbb/install/module_base.php (hunk 9 above)
Comment (zlavdinov) [inline phpBB/phpbb/install/module_base.php]: I think this is the example https://github.com/phpbb/phpbb/blob/master/phpBB/phpbb/install/module_base.php#L123
Initially, I used task index as you said. However, the user could skip over a step without noticing it. Thus, I had to return him the data about which tasks are already completed.
``` $iterator->current()->get_navigation_stage_path() ```

Ok, I will try to retrieve the previously submitted infomation.

File: phpBB/phpbb/install/module_base.php (hunk 12 above)
Comment (zlavdinov) [inline phpBB/phpbb/install/module_base.php]: Yes, but not every subsequent item becomes inactive. For example, if a user is on step 4 and clicked on step 1. Step 2 can still be filled.

File: phpBB/phpbb/install/module_base.php (hunk 9 above)
Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: I don't really think it is necessary to be able to complete these steps in any order, as if i'm understanding correctly it is possible to go forward and backward with your modification.

Also, it would be probably fine if you just let the user jump back to the the part where the install starts. However, in that case as well we need to clear stuff out from the database etc. It is not the simplest issue, that's for sure.

File: phpBB/phpbb/install/module_base.php (hunk 12 above)
Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: Then it is even more simple, as you only need to change to current active item to not active, and the previously filled item to active.

File: phpBB/phpbb/install/module_base.php (hunk 13 above)
Comment (zlavdinov) [inline phpBB/phpbb/install/module_base.php]: Yes, but why would it be set to an invalid value? The system is set by the administrator and he can change files anyways.

File: phpBB/phpbb/install/module_base.php (hunk 12 above)
Comment (zlavdinov) [inline phpBB/phpbb/install/module_base.php]: Ok, I will try to fix it.

File: phpBB/phpbb/install/module_base.php (hunk 11 above)
Comment (zlavdinov) [inline phpBB/phpbb/install/module_base.php]: Yes, I totally agree. I will try to change it.

File: phpBB/phpbb/install/module_base.php (hunk 13 above)
Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: True, but i have another question: then why would you want to add it as a feature? :)

Anyways, the point is that you can still do this client side (e.g. there could obviously be only one active task at a time, so when you send the active menu data before running the task, you can set this from js). 

I think you should move this logic to the front end code, and only modify this code when some backend logic is needed. Btw if you have any questions about the installer, just leave a comment or find me on IRC.

Comment (zlavdinov) [inline phpBB/phpbb/install/module_base.php]: I am sorry, can we start from the beginning?
Do I understand correctly that you are cautions that the administrator can intentionally send incorrect data (task_index) during the installation? 

I do agree that a big part of this solution can and should be moved to client side, something what I will try to do.

Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: Nope, I'm just not sure what you wanted to achieve here.

Comment (zlavdinov) [inline phpBB/phpbb/install/module_base.php]: Yes, actually, I was incorrect and this loop is unnecessary. 
By the way, what should the program do when incorrect value is received?

Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: As long as it couldn't happen without tampering with the JS and/or requests, it really doesn't matter, we can just let it crash IMO.

**Comment by zlavdinov** [comment]
@CHItA could you please review the updates? I tried to fix most of the issues that you commented on. Could you say what is left to be fixed to prepare this PR for merge? 
P.S. Please never mind the commit style for now - I will fix them before merge.

File: phpBB/phpbb/install/module_base.php [hunk 14]
```diff
@@ -108,9 +108,27 @@ public function run()
 		$task_index	= $this->recover_progress();
 		$iterator	= $this->task_collection->getIterator();
 
+
 		if ($task_index < $iterator->count())
 		{
-			$iterator->seek($task_index);
+		    // Receiving task index
```
Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: This should probably be moved to the `obtain_data/install_module.php`.

File: phpBB/phpbb/install/module_base.php [hunk 15]
```diff
@@ -108,9 +108,27 @@ public function run()
...
+		    $get_tast_index = (int)$this->iohandler->get_input('task_index', 0);
+		    // Validation check
+		    if($get_tast_index > 0 && $get_tast_index < $iterator->count()){
+                $get_tast_index--;
+                $iterator->seek($get_tast_index);
+                $this->install_config->set_finished_task($get_tast_index);
+            }
+            else{
+
+                for($i = 0; $i < $task_index; $i++){
```
Comment (CHItA) [inline phpBB/phpbb/install/module_base.php]: This logic probably should move to JS.

File: phpBB/phpbb/install/module_base.php [hunk 16]
```diff
@@ -102,15 +102,35 @@ public function check_requirements()
...
 		// Recover install progress
 		$task_index	= $this->recover_progress();
 		$iterator	= $this->task_collection->getIterator();
 
+
 		if ($task_index < $iterator->count())
 		{
-			$iterator->seek($task_index);
+			// Receiving task index
+			$get_task_index = (int)$this->select_task_index();
```
Comment (zlavdinov) [inline phpBB/phpbb/install/module_base.php]: @CHItA could you say if this is the right way to move `receiving task index` to _install_module.php_
I created the function `select_task_index` at class _module_base_. Then I redefined it at _obtain_data/install_module.php_

**Comment by zlavdinov** [comment]
@CHItA Could you please take a look at the update?
I also have a specific question at line [#123](https://github.com/phpbb/phpbb/pull/4739/files#diff-e36e2b7b4e9d4e2303234ed604175856R123)

**Comment by DavidIQ** [comment]
Commit messages here are wrong and even new commits added after this issue was previously mentioned are also wrong.

**Comment by zlavdinov** [comment]
@DavidIQ yes, thank you for pointing that out. As I [mentioned](https://github.com/phpbb/phpbb/pull/4739#issuecomment-287624628), I intended to fix them after the code will be corrected. I thought that it would be an easier approach for me. Is this ok?

File: phpBB/adm/style/installer_form.html [hunk 17]
```diff
@@ -17,13 +17,13 @@
...
-				<input type="text" name="{options.KEY}" value="{options.DEFAULT}" />
+				<input type="text" name="{options.KEY}" value="<!-- IF options.LOAD -->{options.LOAD}<!-- ELSE -->{options.DEFAULT}<!-- ENDIF -->" />
 				<!-- ENDIF -->
 				<!-- IF options.TYPE == 'email' -->
-				<input type="email" name="{options.KEY}" value="{options.DEFAULT}" />
+				<input type="email" name="{options.KEY}" value="<!-- IF options.LOAD -->{options.LOAD}<!-- ELSE -->{options.DEFAULT}<!-- ENDIF -->" />
 				<!-- ENDIF -->
 				<!-- IF options.TYPE == 'password' -->
-				<input type="password" name="{options.KEY}" />
+				<input type="password" name="{options.KEY}" value="" />
```
Comment (marc1706) [inline phpBB/adm/style/installer_form.html]: Value is optional in HTML5 for anything apart from checkbox and radio elements. There is no need to add it here.

File: phpBB/assets/javascript/installer.js [hunk 18]
```diff
@@ -21,6 +21,26 @@
 	// Intercept form submits
 	interceptFormSubmit($('#install_install'));
 
+	// Interception of click on the navigation link
+	$('.my_send').click( function (e) {
```
Comment (marc1706) [inline phpBB/assets/javascript/installer.js]: I'd recommend using a more meaningful class name that is based on it's purpose. `my_send` actually looks like you're trying to use hungarian notation which is something we do not approve of.

File: phpBB/assets/javascript/installer.js [hunk 19]
```diff
@@ -21,6 +21,26 @@
...
+		e.preventDefault();
+		// check if the previous fields are filled
+		if($(this).parent().hasClass('filled')){
```
Comment (marc1706) [inline phpBB/assets/javascript/installer.js]: Missing space between closing parenthesis and opening curly brace.

File: phpBB/assets/javascript/installer.js [hunk 20]
```diff
@@ -165,11 +186,21 @@
 			$stage = $('#' + navID);
 			$stageListItem = $stage.parent();
 
+			var list = $stageListItem.parent().children();
+			for (var i = 0; i < list.length; i++){
```
Comment (marc1706) [inline phpBB/assets/javascript/installer.js]: Missing space as mentioned above.

File: phpBB/assets/javascript/installer.js [hunk 21]
```diff
@@ -165,11 +186,21 @@
...
+				if(list[i] == $stageListItem.get(0))
+					break;
```
Comment (marc1706) [inline phpBB/assets/javascript/installer.js]: We don't omit parentheses.

File: phpBB/assets/javascript/installer.js [hunk 22]
```diff
@@ -165,11 +186,21 @@
...
+				$(list[i]).children().addClass('filled');
+			}
+			$stage.addClass('filled');
+
 			if ($active.length && !$active.is($stageListItem)) {
 				$active.removeAttr('id');
 			}
 
 			$stageListItem.attr('id', 'activemenu');
+
```
Comment (marc1706) [inline phpBB/assets/javascript/installer.js]: Please remove this not needed empty new line.

File: phpBB/phpbb/install/controller/helper.php [hunk 23]
```diff
@@ -311,11 +311,16 @@ protected function render_navigation()
 				{
 					if (isset($sub_entry['stage']) && $sub_entry['stage'] === true)
 					{
+						// sending parameters for navigation link
 						$this->template->assign_block_vars('l_block2', array(
 							'L_TITLE' => $this->language->lang($sub_entry['label']),
 							'S_SELECTED' => (isset($sub_entry['selected']) && $sub_entry['selected'] === true),
 							'S_COMPLETE' => (isset($sub_entry['completed']) && $sub_entry['completed'] === true),
 							'STAGE_NAME' => $name,
+                            'IS_LINK' => (isset($sub_entry['is_link']) && $sub_entry['is_link'] === true),
```
Comment (marc1706) [inline phpBB/phpbb/install/controller/helper.php]: We use tabs instead of spaces as stated in our coding guidelines.

File: phpBB/phpbb/install/helper/iohandler/ajax_iohandler.php [hunk 24]
```diff
@@ -186,6 +186,7 @@ public function generate_form_render_data($title, $form)
 			$tpl_ary['TITLE'] = $this->language->lang($input_options['label']);
 			$tpl_ary['KEY'] = $input_name;
 			$tpl_ary['S_EXPLAIN'] = false;
+            $tpl_ary['LOAD'] = isset($input_options['load']) ? $input_options['load'] : false;
```
Comment (marc1706) [inline phpBB/phpbb/install/helper/iohandler/ajax_iohandler.php]: Same tabs vs. spaces issues as stated above.

File: phpBB/phpbb/install/helper/navigation/install_navigation.php [hunk 25]
```diff
@@ -58,15 +58,55 @@ public function get()
 						'stage'	=> true,
 						'order'	=> 1,
 					),
-					'obtain_data'	=> array(
-						'label'	=> 'STAGE_OBTAIN_DATA',
+					'obtain_admin_data'	=> array(
+						'label'	=> 'ADMIN_CONFIG',
 						'stage'	=> true,
+                        'is_link'	=> true,
```
Comment (marc1706) [inline phpBB/phpbb/install/helper/navigation/install_navigation.php]: Tabs instead of spaces.

File: phpBB/phpbb/install/helper/navigation/install_navigation.php [hunk 26]
```diff
@@ -58,15 +58,55 @@ public function get()
...
+                        'link'	=> '/install/app.php/install',
```
Comment (marc1706) [inline phpBB/phpbb/install/helper/navigation/install_navigation.php]: I don't think we should use hardcoded links in here. That's what we have the routing system for.

File: phpBB/phpbb/install/installer.php [hunk 27]
```diff
@@ -13,6 +13,7 @@
 
 namespace phpbb\install;
 
+use GuzzleHttp\Message\Request;
```
Comment (marc1706) [inline phpBB/phpbb/install/installer.php]: You're adding new `use` statements but you're not changing anything in this file. Please remove any unnecessary changes.

File: phpBB/phpbb/install/installer.php [hunk 28]
```diff
@@ -26,6 +27,7 @@
 use phpbb\install\helper\iohandler\cli_iohandler;
 use phpbb\install\helper\iohandler\iohandler_interface;
 use phpbb\path_helper;
+use phpbb\request\request_interface;
```
Comment (marc1706) [inline phpBB/phpbb/install/installer.php]: Same as stated above.

File: phpBB/phpbb/install/module/obtain_data/install_module.php [hunk 29]
```diff
@@ -30,4 +30,12 @@ public function get_step_count()
 	{
 		return 0;
 	}
+
+	/**
+	 * {@inheritdoc}
```
Comment (marc1706) [inline phpBB/phpbb/install/module/obtain_data/install_module.php]: Where is it inheriting from? I don't see any interface definition for this?

File: phpBB/phpbb/install/module/obtain_data/install_module.php [hunk 30]
```diff
@@ -30,4 +30,12 @@ public function get_step_count()
...
+	 */
+	public function select_task_index()
+	{
+		return $get_tast_index = (int)$this->iohandler->get_input('task_index', 0);
```
Comment (marc1706) [inline phpBB/phpbb/install/module/obtain_data/install_module.php]: Space between `(int)` and `$this->...` missing.

File: phpBB/phpbb/install/module/obtain_data/task/obtain_admin_data.php [hunk 31]
```diff
@@ -112,20 +112,24 @@ protected function request_form_data($use_request_data = false)
 				'description'	=> 'ADMIN_USERNAME_EXPLAIN',
 				'type'			=> 'text',
 				'default'		=> $admin_username,
+                'load'		=> $this->install_config->get('admin_name')
```
Comment (marc1706) [inline phpBB/phpbb/install/module/obtain_data/task/obtain_admin_data.php]: More spaces instead of tabs. Please review your complete code for this.

File: phpBB/phpbb/install/module/obtain_data/task/obtain_admin_data.php [hunk 32]
```diff
@@ -215,4 +219,10 @@ public function get_task_lang_name()
 	{
 		return '';
 	}
+
+	// Returns the task name for navigation bar
```
Comment (marc1706) [inline phpBB/phpbb/install/module/obtain_data/task/obtain_admin_data.php]: Please add a proper docblock as employed by other methods. It should state what it does and what it's returning in this case.

File: phpBB/phpbb/install/module/obtain_data/task/obtain_board_data.php [hunk 33]
```diff
@@ -136,7 +136,7 @@ protected function request_form_data($use_request_data = false)
 			$lang_options[] = array(
 				'value'		=> $lang['iso'],
 				'label'		=> $lang['local_name'],
-				'selected'	=> ($default_lang === $lang['iso']),
+				'selected'	=> ((($default_lang === $lang['iso']) && !$this->install_config->get('board_name')) || ($this->install_config->get('board_name') == $lang['iso']) ),
```
Comment (marc1706) [inline phpBB/phpbb/install/module/obtain_data/task/obtain_board_data.php]: Consider splitting this into multiple lines.

File: phpBB/phpbb/install/module/obtain_data/task/obtain_board_data.php [hunk 34]
```diff
@@ -182,4 +184,10 @@ public function get_task_lang_name()
 	{
 		return '';
 	}
+
+	// Returns the task name for navigation bar
+    public function get_navigation_stage_path()
```
Comment (marc1706) [inline phpBB/phpbb/install/module/obtain_data/task/obtain_board_data.php]: Same issue as stated above.

File: phpBB/phpbb/install/module/obtain_data/task/obtain_database_data.php [hunk 35]
```diff
@@ -267,4 +273,10 @@ public function get_task_lang_name()
 	{
 		return '';
 	}
+
+	// Returns the task name for navigation bar
+    public function get_navigation_stage_path()
```
Comment (marc1706) [inline phpBB/phpbb/install/module/obtain_data/task/obtain_database_data.php]: Same issue as stated above.

File: phpBB/phpbb/install/module/obtain_data/task/obtain_email_data.php [hunk 36]
```diff
@@ -107,41 +107,46 @@ public function run()
...
-							'selected'	=> false,
+							'selected'	=> $this->install_config->get('smtp_delivery'),
 						),
 					),
 				),
 				'smtp_host' => array(
 					'label'			=> 'SMTP_SERVER',
 					'type'			=> 'text',
 					'default'		=> $smtp_host,
+                    'load'		=> $this->install_config->get('smtp_host')
```
Comment (marc1706) [inline phpBB/phpbb/install/module/obtain_data/task/obtain_email_data.php]: Indentation is completely off.

File: phpBB/phpbb/install/module/obtain_data/task/obtain_email_data.php [hunk 37]
```diff
@@ -170,4 +175,10 @@ public function get_task_lang_name()
 	{
 		return '';
 	}
+
+	// Returns the task name for navigation bar
```
Comment (marc1706) [inline phpBB/phpbb/install/module/obtain_data/task/obtain_email_data.php]: More indentation problems and missing docblock.

File: phpBB/phpbb/install/module_base.php [hunk 38]
```diff
@@ -102,15 +102,35 @@ public function check_requirements()
 	/**
 	 * {@inheritdoc}
 	 */
+	public function select_task_index()
```
Comment (marc1706) [inline phpBB/phpbb/install/module_base.php]: Again, where is it inheriting the docblock from?

File: phpBB/phpbb/install/module_base.php [hunk 39]
```diff
@@ -102,15 +102,35 @@ public function check_requirements()
...
 		$iterator	= $this->task_collection->getIterator();
 
+
 		if ($task_index < $iterator->count())
 		{
-			$iterator->seek($task_index);
+			// Receiving task index
+			$get_task_index = (int)$this->select_task_index();
+			// Validation check
+			if($get_task_index)
```
Comment (marc1706) [inline phpBB/phpbb/install/module_base.php]: Missing space after if.

File: phpBB/phpbb/install/module_base.php [hunk 40]
```diff
@@ -102,15 +102,35 @@ public function check_requirements()
...
+			{
+				$get_task_index--;
+				$iterator->seek($get_task_index);
+				$this->install_config->set_finished_task($get_task_index);
+			}
+			else{
```
Comment (marc1706) [inline phpBB/phpbb/install/module_base.php]: Curly brace on new line.

File: phpBB/phpbb/install/task_base.php [hunk 41]
```diff
@@ -50,4 +50,10 @@ public function check_requirements()
 	{
 		return true;
 	}
+
+	// Returns the task name for navigation bar
+    public function get_navigation_stage_path()
```
Comment (marc1706) [inline phpBB/phpbb/install/task_base.php]: Same issues as stated before.

**Comment by zlavdinov** [comment]
@CHItA @marc1706 Could you take a look at the latest changes?

**Comment by Nicofuma** [comment]
@zlavdinov @CHItA @marc1706 
Any news on this PR?

File: phpBB/phpbb/install/task_interface.php [hunk 42]
```diff
@@ -58,4 +58,11 @@ public function run();
 	 * @return string
 	 */
 	public function get_task_lang_name();
+	
+	/**
+	 * Returns an array to the correct navigation stage
+	 *
+	 * @return array
+	 */
+    public function get_navigation_stage_path()
```
Comment (marc1706) [inline phpBB/phpbb/install/task_interface.php]: Also, use tabs instead of spaces.

**Comment by CHItA** [comment]
closing this due to the author's inactivity. Feel free to reopen it if you wish to continue working on it.
